ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -1152,42 +1147,62 @@ private void encryptChunk() {
-            }));
+                isMetadataChunk() ? null : toEncrypt.retainedDuplicate(), ByteBuffer.wrap(chunkUserMetadata),
+                kms.getRandomKey(), cryptoService, kms, encryptJobMetricsTracker,
+                (EncryptJob.EncryptJobResult result, Exception exception) -> {
+                  logger.trace("Processing encrypt job callback for chunk at index {}", chunkIndex);
+                  if (!isMetadataChunk()) {
+                    releaseBlobContent();
+                  }
+                  encryptJobMetricsTracker.onJobResultProcessingStart();
+                  if (exception == null && !isOperationComplete()) {
+                    if (!isMetadataChunk()) {
+                      buf = result.getEncryptedBlobContent();
+                    }
+                    encryptedPerBlobKey = result.getEncryptedKey();
+                    chunkUserMetadata = result.getEncryptedUserMetadata().array();
+                    logger.trace("Completing encrypt job result for chunk at index {}", chunkIndex);
+                    prepareForSending();
+                    chunkReadyAtMs = time.milliseconds();
+                  } else {
+                    encryptJobMetricsTracker.incrementOperationError();
+                    if (!isOperationComplete()) {
+                      logger.trace("Setting exception from encrypt of chunk at index {} ", chunkIndex, exception);
+                      // If we are here, then the result is null. no need to release it.
+                      setOperationExceptionAndComplete(new RouterException(
+                          "Exception thrown on encrypting the content for chunk at index " + chunkIndex, exception,
+                          RouterErrorCode.UnexpectedInternalError));
+                    } else {
+                      logger.trace(
+                          "Ignoring exception from encrypt job for chunk at index {} as operation exception {} is set already",
+                          chunkIndex, getOperationException(), exception);
+                      // If we are here, then the operation is completed and the exception could be null, in this case,
+                      // we have to release the content in the result.
+                      if (result != null) {
+                        result.release();
+                      }
+                    }
+                  }
+                  routerMetrics.encryptTimeMs.update(time.milliseconds() - chunkEncryptReadyAtMs);
+                  encryptJobMetricsTracker.onJobResultProcessingComplete();
+                  routerCallback.onPollReady();
+                  // double check if the operation is not completed. If so, we have to release the buf here, since in
+                  // main thread, chunk might already be released.
+                  if (isOperationComplete()) {
+                    logger.info("Clear put chunk in encryption callback since operation is completed");
+                    clear();
```

Comment:
The clear() method does not seem like it was meant to be called from non-main threads. Is it necessary to do all of the cleanup or just call `releaseBlobContent` here and in `fillChunks()`?

Reply:
make sense